File: server/bazelrunner/src/main/java/org/jetbrains/bsp/bazel/bazelrunner/BazelProcess.java
package org.jetbrains.bsp.bazel.bazelrunner;

import io.vavr.control.Option;
import java.io.InputStream;
import java.util.List;
import org.jetbrains.bsp.bazel.bazelrunner.data.BazelProcessResult;
import org.jetbrains.bsp.bazel.bazelrunner.utils.BazelStreamReader;
import org.jetbrains.bsp.bazel.server.loggers.BuildClientLogger;

public class BazelProcess {

  private static final int OK_EXIT_CODE = 0;

  private final Process process;
  private final Option<BuildClientLogger> buildClientLogger;

  BazelProcess(Process process, Option<BuildClientLogger> buildClientLogger) {
    this.process = process;
    this.buildClientLogger = buildClientLogger;
  }

  public BazelProcessResult waitAndGetResult() {
    try {
      int exitCode = process.waitFor();
      BazelProcessResult bazelProcessResult =
          new BazelProcessResult(process.getInputStream(), process.getErrorStream(), exitCode);

      logBazelMessage(bazelProcessResult, exitCode);

      return bazelProcessResult;
    } catch (InterruptedException e) {
      throw new RuntimeException(e);
    }
  }

  private void logBazelMessage(BazelProcessResult bazelProcessResult, int exitCode) {
    if (exitCode == OK_EXIT_CODE) {
      logBazelMessage(bazelProcessResult);
    } else {
      logBazelError(bazelProcessResult);
    }
  }

  private void logBazelMessage(BazelProcessResult bazelProcessResult) {
    String message = bazelProcessResult.getStderr();
    buildClientLogger.forEach(logger -> logger.logMessage(message));
  }

  private void logBazelError(BazelProcessResult bazelProcessResult) {
    String error = bazelProcessResult.getStderr();
    buildClientLogger.forEach(logger -> logger.logError(error));
  }

  public InputStream getInputStream() {
    return process.getInputStream();
  }

  public List<String> getStderr() {
    return BazelStreamReader.drainStream(process.getErrorStream());
  }
}

File: server/bazelrunner/src/main/kotlin/org/jetbrains/bsp/bazel/bazelrunner/BazelProcess.kt
package org.jetbrains.bsp.bazel.bazelrunner

import org.apache.logging.log4j.LogManager
import org.eclipse.lsp4j.jsonrpc.CancelChecker
import org.jetbrains.bsp.bazel.bazelrunner.outputs.AsyncOutputProcessor
import org.jetbrains.bsp.bazel.bazelrunner.outputs.OutputProcessor
import org.jetbrains.bsp.bazel.bazelrunner.outputs.SyncOutputProcessor
import org.jetbrains.bsp.bazel.commons.BazelStatus
import org.jetbrains.bsp.bazel.commons.Format
import org.jetbrains.bsp.bazel.commons.Stopwatch
import org.jetbrains.bsp.bazel.logger.BspClientLogger
import java.time.Duration
import java.util.concurrent.CompletableFuture

class BazelProcess internal constructor(
  val process: Process,
  private val logger: BspClientLogger? = null,
  private val serverPidFuture: CompletableFuture<Long>?,
  private val finishCallback: () -> Unit = {},
) {
  fun waitAndGetResult(cancelChecker: CancelChecker, ensureAllOutputRead: Boolean = false): BazelProcessResult {
    return try {
      val stopwatch = Stopwatch.start()
      val outputProcessor: OutputProcessor =
        if (logger != null) {
          if (ensureAllOutputRead) {
            SyncOutputProcessor(process, logger::message, LOGGER::info)
          } else {
            AsyncOutputProcessor(process, logger::message, LOGGER::info)
          }
        } else {
          if (ensureAllOutputRead) {
            SyncOutputProcessor(process, LOGGER::info)
          } else {
            AsyncOutputProcessor(process, LOGGER::info)
          }
        }

      val exitCode = outputProcessor.waitForExit(cancelChecker, serverPidFuture, logger)
      val duration = stopwatch.stop()
      logCompletion(exitCode, duration)
      return BazelProcessResult(outputProcessor.stdoutCollector, outputProcessor.stderrCollector, BazelStatus.fromExitCode(exitCode))
    } finally {
      finishCallback()
    }
  }

  private fun logCompletion(exitCode: Int, duration: Duration) {
    logger?.message("Command completed in %s (exit code %d)", Format.duration(duration), exitCode)
  }

  companion object {
    private val LOGGER = LogManager.getLogger(BazelProcess::class.java)
  }
}
